### DISCUSSION TITLE
Questions about next/image caching in the Next.js Pages Router

### DISCUSSION THREAD

**Discussion by shape1541** [vercel/next.js#70143] [Help]
### Summary

I have a question regarding the `next/image` component in the Next.js Pages Router. Although I am using Next.js 13, I assume the behavior is similar or the same in the latest version, Next.js 14.

According to the official documentation for `next/image` in the Pages Router (https://nextjs.org/docs/pages/api-reference/components/image), the `minimumCacheTTL` property can be set like this:
```javascript
module.exports = {
  images: {
    minimumCacheTTL: 60,
  },
};
```

I’m curious about the criteria for determining the appropriate minimumCacheTTL value. For instance, if images in a community post or e-commerce product are updated frequently, and a user revisits the page before the cache duration expires, will they still see the old image because the cached version is still valid?

Additionally, would this be an issue only if the image URL remains the same? If the image’s URL changes after an update, does it bypass the cache and display the new image? How can I handle these scenarios efficiently?

**Comment by jattu8602** [comment]
From my understanding, `minimumCacheTTL` controls the minimum amount of time that Next.js Image Optimization can cache an optimized image before it needs to revalidate or fetch a fresh version.

### Scenario 1: Image content changes, but the URL remains the same

Suppose you have:

```text
https://example.com/product.jpg
```

If the image is updated on the server but the URL stays unchanged, users may continue seeing the cached version until the cache expires.

For example:

```js
images: {
  minimumCacheTTL: 3600, // 1 hour
}
```

If a user visits the page, the optimized image may be cached for up to 1 hour. If the image content changes during that period, users could still receive the older cached version until the cache is refreshed.

### Scenario 2: Image URL changes after an update

If the URL changes, such as:

```text
https://example.com/product.jpg?v=1
```

becoming

```text
https://example.com/product.jpg?v=2
```

or

```text
https://example.com/product-v2.jpg
```

then browsers, CDNs, and Next.js treat it as a completely new resource.

In this case, users will receive the updated image immediately because the cache key is different.

### Recommended approach for frequently updated images

For applications such as:

* Community posts
* User profile pictures
* Product catalogs
* CMS-managed content

it is generally recommended to use cache-busting/versioned URLs rather than relying on a very low `minimumCacheTTL`.

Examples:

```text
/avatar.jpg?v=123
/product.jpg?v=20240916
/image.abc123.jpg
```

This allows you to:

* Keep longer cache durations for better performance
* Reduce bandwidth usage
* Ensure users receive updated images whenever the content changes

### How I usually think about `minimumCacheTTL`

* Images rarely change → use a higher value.
* Images change frequently but URLs are versioned → higher value is still fine.
* Images change frequently and URLs remain identical → consider a lower value or implement URL versioning.

So, if I understand correctly, the primary concern is not the cache duration itself, but whether the image URL changes when the image content changes. Using versioned URLs appears to be the most reliable strategy for balancing performance and freshness.

**Comment by mohibali07** [comment]
Hi @shape1541, great questions regarding `next/image` caching! Here is a breakdown of how the behavior works in these scenarios:

**1. Will users see the old image if revisited before the cache expires?**
Yes. If the `minimumCacheTTL` duration hasn't passed and the image URL remains exactly the same, the Next.js server (or your edge network/CDN) will continue to serve the cached, optimized version. The user will see the stale image until the TTL expires and a new request triggers a re-optimization in the background.

**2. Does changing the URL bypass the cache?**
Exactly! Next.js determines the cache key based on a combination of the image's `src`, width, quality, and output format. If the `src` URL changes, Next.js treats it as a completely new and unique image request. It will instantly bypass the old cache, fetch the new image, optimize it, and start a new cache cycle for that specific URL.

**3. How to handle this efficiently (E-commerce / Community Posts)?**
*   **Use Cache Busting / Content Hashing (Best Practice):** Instead of keeping the exact same filename for an updated image (e.g., `product.jpg`), append a hash, version, or timestamp to the image URL whenever it gets updated in your database (e.g., `product.jpg?v=1726500000` or `product-v2.jpg`). This forces an immediate cache bypass without relying on the TTL. Modern headless CMS platforms (like Sanity.io) often handle this automatically by appending unique IDs or hashes to their image CDN URLs.
*   **Optimize the `minimumCacheTTL`:** Setting the TTL too low (e.g., 1 second) will cause your server to constantly re-optimize images, which heavily increases CPU load and server costs. The default 60 seconds is usually a great balance. Rely on URL changes (cache busting) for instant updates, and let the TTL protect your server from redundant processing during high traffic.

Hope this clarifies the behavior for your project!
